slang.h

```diff
@@ -2543,8 +2597,329 @@ namespace slang
+
+            /** Specialize a type based on type arguments.
+            */
+        virtual SLANG_NO_THROW TypeReflection* SLANG_MCALL specializeType(
+            TypeReflection*             type,
+            SlangInt                    specializationArgCount,
+            SpecializationArg const*    specializationArgs,
+            ISlangBlob**                outDiagnostics = nullptr) = 0;
+
+
+            /** Get the layout `type` on the chosen `target`.
+            */
+        virtual SLANG_NO_THROW TypeLayoutReflection* SLANG_MCALL getTypeLayout(
+            TypeReflection* type,
+            SlangInt        targetIndex = 0,
+            LayoutRules     rules = LayoutRules::Default,
+            ISlangBlob**    outDiagnostics = nullptr) = 0;
+
+            /** Create a request to load/compile front-end code.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL createCompileRequest(
+            SlangCompileRequest**   outCompileRequest) = 0;
+
+#if 0
+            /** Add a path to use when searching for `#include`d or `import`ed files.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL addSearchPath(
+            char const* path) = 0;
+
+            /** Add a global preprocessor definition to use for all compilation.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL addPreprocessorDefine(
+            char const* name,
+            char const* value) = 0;
+
+            /** Set the default matrix layout mode to use for all compiles.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL setMatrixLayoutMode(
+            SlangMatrixLayoutMode mode) = 0;
+#endif
+    };
+
+    #define SLANG_UUID_ISession { 0x67618701, 0xd116, 0x468f, { 0xab, 0x3b, 0x47, 0x4b, 0xed, 0xce, 0xe, 0x3d } }
+
+
+        /** A module is the granularity of shader code compilation and loading.
+
+        In most cases a module corresponds to a single compile "translation unit."
+        This will often be a single `.slang` or `.hlsl` file and everything it
+        `#include`s.
+
+        Notably, a module `M` does *not* include the things it `import`s, as these
+        as distinct modules that `M` depends on. There is a directed graph of
+        module dependencies, and all modules in the graph must belong to the
+        same session (`ISession`).
+        */
+    struct IModule : public ISlangUnknown
+    {
+    public:
+    };
+    
+    #define SLANG_UUID_IModule { 0xc720e64, 0x8722, 0x4d31, { 0x89, 0x90, 0x63, 0x8a, 0x98, 0xb1, 0xc2, 0x79 } }
+
+
+        /** Argument used for specialization to types/values.
+        */
+    struct SpecializationArg
+    {
+        enum class Kind : int32_t
+        {
+            Unknown,
+            Type,
+        };
+        Kind kind;
+        union
+        {
+            TypeReflection* type;
+        };
+    };
+
+        /** Description of a program to be created.
+        */
+    struct ProgramDesc
+    {
+        struct Item
```

Comment:
Just note that there is no binary backwards compatibility in Desc'y types.

Reply:
Yes, I haven't added the hooks for backwards compatibility here yet, and there is *no* intention to support the API in this PR as-is in the long run (this is a milestone toward the first COM-lite API that will provide backwards compatibility, but it is not yet that point).

I'm still on the fence about whether to do the `sizeof` trick or Vulkan-style self-identifying structure types with support for extensions. My gut wants to do the latter (so that seldom-used features don't clutter up the structures for ordinary users), but my brain says that is probably over-engineering things.